### PR TITLE
[10.x] Make the `updateOrCreate` methods in relations use `firstOrCreate` behind the scenes

### PR DESCRIPTION
**Pull request by mpyw**
Related to:

- #48160
- #48192

Add consistent fix for `updateOrCreate` based on #48160, since there was a consistent fix for `firstOrCreate` in #48192.

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php [hunk 1]
```diff
@@ -672,19 +672,13 @@ public function createOrFirst(array $attributes = [], array $values = [], array
...
-        if (is_null($instance = (clone $this)->where($attributes)->first())) {
-            if (is_null($instance = $this->related->where($attributes)->first())) {
-                return $this->create(array_merge($attributes, $values), $joining, $touch);
-            } else {
-                $this->attach($instance, $joining, $touch);
-            }
-        }
-
-        $instance->fill($values);
+        return tap($this->firstOrCreate($attributes, $values), function ($instance) use ($values) {
```
Comment (taylorotwell) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php:675]: But you're not even using the `$joining` attribute? We need to be really careful we aren't just breaking stuff here.

Comment (mpyw) [reply src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php:675]: Nice catch! I'll fix it

File: src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php [hunk 2]
```diff
@@ -672,11 +672,11 @@ public function createOrFirst(array $attributes = [], array $values = [], array
      */
     public function updateOrCreate(array $attributes, array $values = [], array $joining = [], $touch = true)
     {
-        return tap($this->firstOrCreate($attributes, $values), function ($instance) use ($values) {
+        return tap($this->firstOrCreate($attributes, $values, $joining, $touch), function ($instance) use ($values, $touch) {
             if (! $instance->wasRecentlyCreated) {
                 $instance->fill($values);
 
-                $instance->save(['touch' => false]);
+                $instance->save(compact('touch'));
```
Comment (mpyw) [inline src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php:679]: Is it safe to pass the `$touch` argument to both `firstOrCreate` and `save`? If the timestamp update process runs twice, how should we handle it...

Comment (mpyw) [reply src/Illuminate/Database/Eloquent/Relations/BelongsToMany.php]: That was a mistake, okay

**Pull request merged**